llvm/labs/lab2/kulagin_aleksandr/magic_inline.cpp

```diff
@@ -0,0 +1,116 @@
+#include "llvm/IR/BasicBlock.h"
+#include "llvm/IR/DerivedTypes.h"
+#include "llvm/IR/Function.h"
+#include "llvm/IR/Instruction.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/LLVMContext.h"
+#include "llvm/IR/Value.h"
+#include "llvm/Passes/PassBuilder.h"
+#include "llvm/Passes/PassPlugin.h"
+#include "llvm/Support/Casting.h"
+#include "llvm/Transforms/Utils/BasicBlockUtils.h"
+#include "llvm/Transforms/Utils/Cloning.h"
+#include "llvm/Transforms/Utils/ValueMapper.h"
+#include <map>
+#include <vector>
+
+namespace {
+llvm::BasicBlock *splitBlockBefore(llvm::BasicBlock *Old,
+                                   llvm::Instruction *SplitPt,
+                                   llvm::DomTreeUpdater *DTU = nullptr,
+                                   llvm::LoopInfo *LI = nullptr,
+                                   llvm::MemorySSAUpdater *MSSAU = nullptr,
+                                   const llvm::Twine &BBName = "",
+                                   bool Before = true) {
+  return llvm::SplitBlock(Old, SplitPt, DTU, LI, MSSAU, BBName, Before);
+}
+
+class KulaginMagicInlinePass
+    : public llvm::PassInfoMixin<KulaginMagicInlinePass> {
+public:
+  llvm::PreservedAnalyses run(llvm::Function &Func,
+                              llvm::FunctionAnalysisManager &) {
+    std::vector<llvm::Instruction *> InstrDeleteList;
+    llvm::LLVMContext &CTX = Func.getContext();
+    for (llvm::BasicBlock &Block : Func) {
+      for (llvm::Instruction &Instr : Block) {
+        if (llvm::CallInst *CallInstruction =
+                llvm::dyn_cast<llvm::CallInst>(&Instr)) {
```

Comment:
In LLVM projects earlier check and exit from a loop or other compound statement is usually used to reduce blocks nesting in a code.
```suggestion
        auto *CallInstruction = llvm::dyn_cast<llvm::CallInst>(&Instr);
        if (!CallInstruction) {
          continue;
        }
```